File: src/main/ruby/diffj.rb
#!/usr/bin/jruby -w
# -*- ruby -*-

require 'rubygems'
require 'java'
require 'riel'
require 'diffj/io'
require 'diffj/analysis/short_report'
require 'diffj/analysis/long_report'
require 'diffj/io/diffwriter/context'
require 'diffj/io/diffwriter/context_highlight'
require 'diffj/io/diffwriter/no_context'

include Java

java_import org.incava.diffj.Options
java_import org.incava.diffj.DiffJException
java_import org.incava.analysis.BriefReport
java_import org.incava.analysis.DetailedReport

Log::level = Log::DEBUG
Log.set_widths(-15, 5, -50)

module DiffJ
  class CLI
    include Loggable
    
    attr_reader :exit_value
    attr_reader :report
    
    def initialize brief, context, highlight, recurse, from_label, fromver, to_label, tover
      writer = java.io.OutputStreamWriter.new java.lang.System.out
      @report = brief ? DiffJ::Analysis::ShortReport.new(writer) : DiffJ::Analysis::LongReport.new(writer, context, highlight)
      @recurse = recurse
      @from_label = from_label
      @fromver = fromver
      @to_label = to_label
      @tover = tover
      @jef = DiffJ::IO::Factory.new
      @exit_value = 0
    end

    def create_java_element fname, label, source
      begin
        info "fname: #{fname}"
        @jef.create_element java.io.File.new(fname), label, source, @recurse
      rescue StandardError => de
        $stderr.puts de.message
        @exit_value = 1
        nil
      end
    end
    
    def create_to_element to_name
      create_java_element to_name, @to_label, @tover
    end
    
    def create_from_element from_name
      create_java_element from_name, @from_label, @fromver
    end

    def compare from_name, to_elmt
      begin 
        from_elmt = create_from_element from_name
        return false if from_elmt.nil?
        from_elmt.compare_to @report, to_elmt
        @exit_value = @report.differences.was_added? ? 1 : 0
        true
      rescue DiffJException => de
        $stderr.puts de.message
        @exit_value = 1
        nil
      end
    end

    def process_names names
      if names.size < 2
        $stderr.puts "usage: diffj from-file to-file"
        exit_value = 1
        return
      end

      return unless to_elmt = create_to_element(names[-1])
      names[0 ... -1].each do |fromname|
        compare fromname, to_elmt
      end
    end
  end
end

if __FILE__ == $0
  opts = Options.get
  names = opts.process ARGV
  
  diffj = DiffJ::CLI.new(opts.showBriefOutput, 
                         opts.showContextOutput, 
                         opts.highlightOutput,
                         opts.recurse,
                         opts.firstFileName, opts.getFromSource,
                         opts.getSecondFileName, opts.getToSource)
  
  rarray = Array.new
  names.each do |name|
    rarray << name
  end

  diffj.process_names rarray
  puts "exiting with value: #{diffj.exit_value}"
  exit diffj.exit_value
end

File: src/main/java/org/incava/diffj/app/DiffJ.java
package org.incava.diffj.app;

import java.io.File;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.util.Arrays;
import java.util.List;
import org.incava.analysis.BriefReport;
import org.incava.analysis.DetailedReport;
import org.incava.analysis.FileDiffs;
import org.incava.analysis.Report;
import org.incava.diffj.io.JavaElementFactory;
import org.incava.diffj.io.JavaFSElement;
import org.incava.diffj.lang.DiffJException;
import org.incava.ijdk.util.ListExt;

public class DiffJ {
    private final Report report;
    private int exitValue;
    private final boolean recurseDirectories;
    private final String fromLabel;
    private final String toLabel;
    private final String fromSource;
    private final String toSource;
    private final JavaElementFactory jef;
    private final FileDiffs fileDiffs;
    
    public DiffJ(boolean briefOutput, boolean contextOutput, boolean highlightOutput, 
                 boolean recurseDirectories,
                 String fromLabel, String fromSource,
                 String toLabel, String toSource) {
        // tr.Ace.set(true, 25, 4, 20, 25);
        // tr.Ace.setOutput(Log.VERBOSE, Log.LEVEL4);
        // tr.Ace.setOutput(Log.QUIET,   Log.LEVEL2);
        // tr.Ace.setVerbose(true);
        // tr.Ace.stack("this", this, 15);

        Writer writer = new OutputStreamWriter(System.out);
        this.report = briefOutput ? new BriefReport(writer) : new DetailedReport(writer, contextOutput, highlightOutput);
        this.recurseDirectories = recurseDirectories;
        this.fromLabel = fromLabel;
        this.toLabel = toLabel;
        this.fromSource = fromSource;
        this.toSource = toSource;
        this.exitValue = 0;
        this.jef = new JavaElementFactory();
        this.fileDiffs = report.getDifferences();
    }

    protected Report getReport() {
        return report;
    }

    protected int getExitValue() {
        return exitValue;
    }

    protected void setExitValue(int ev) {
        exitValue = ev;
    }

    protected JavaFSElement getJavaElement(File file, String label, String source) {
        try {
            return jef.createElement(file, label, source, recurseDirectories);
        }
        catch (DiffJException de) {
            // de.printStackTrace(System.out);
            System.err.println(de.getMessage());
            exitValue = 1;
            return null;
        }
    }

    public JavaFSElement getToElement(String toName) {
        return getJavaElement(new File(toName), toLabel, toSource);
    }

    public JavaFSElement getFromElement(String fromName) {
        return getJavaElement(new File(fromName), fromLabel, fromSource);
    }

    public boolean compareElements(String fromName, JavaFSElement toElmt) {
        try {
            JavaFSElement fromElmt = getFromElement(fromName);
            if (fromElmt == null) {
                return false;
            }
            fromElmt.compareTo(report, toElmt);
            if (fileDiffs.wasAdded()) {
                exitValue = 1;
            }
            return true;
        }
        catch (DiffJException de) {
            System.err.println(de.getMessage());
            exitValue = 1;
            return false;
        }
    }

    public void processNames(List<String> names) {
        if (names.size() < 2) {
            System.err.println("usage: diffj from-file to-file");
            exitValue = 1;
            return;
        }

        String lastName = ListExt.get(names, -1);
        JavaFSElement toElmt = getToElement(lastName);
        if (toElmt == null) {
            return;
        }

        for (int ni = 0; ni < names.size() - 1; ++ni) {
            if (!compareElements(names.get(ni), toElmt)) {
                break;
            }
        }
        tr.Ace.log("exitValue", "" + exitValue);
    }

    public static void main(String[] args) {
        Options opts = new Options();
        List<String> names = opts.process(Arrays.asList(args));

        if (opts.showVersion()) {
            System.out.println("diffj, version " + Options.VERSION);
            System.out.println("Written by Jeff Pace (jpace [at] incava [dot] org)");
            System.out.println("Released under the Lesser GNU Public License");
            System.exit(0);
        }

        DiffJ diffj = new DiffJ(opts.showBriefOutput(), opts.showContextOutput(), opts.highlightOutput(),
                                opts.recurse(),
                                opts.getFirstFileName(), opts.getFromSource(),
                                opts.getSecondFileName(), opts.getToSource());
        diffj.processNames(names);
        System.exit(diffj.exitValue);
    }
}
